src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java

```diff
@@ -95,174 +59,26 @@ public void addBinaryField(FieldInfo field, DocValuesProducer valuesProducer) th
-        }
-        final String engineName = field.attributes().getOrDefault(KNNConstants.KNN_ENGINE, KNNEngine.DEFAULT.getName());
-        return KNNEngine.getEngine(engineName);
-    }
-
-    public void addKNNBinaryField(FieldInfo field, DocValuesProducer valuesProducer, boolean isMerge, boolean isRefresh)
-        throws IOException {
-        // Get values to be indexed
-        BinaryDocValues values = valuesProducer.getBinary(field);
-        final KNNEngine knnEngine = getKNNEngine(field);
-        final String engineFileName = buildEngineFileName(
-            state.segmentInfo.name,
-            knnEngine.getVersion(),
-            field.name,
-            knnEngine.getExtension()
-        );
-        final String indexPath = Paths.get(
-            ((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(),
-            engineFileName
-        ).toString();
-
-        // Determine if we are creating an index from a model or from scratch
-        NativeIndexCreator indexCreator;
-        KNNCodecUtil.Pair pair;
-        Map<String, String> fieldAttributes = field.attributes();
-        VectorDataType vectorDataType;
-
-        if (fieldAttributes.containsKey(MODEL_ID)) {
-            String modelId = fieldAttributes.get(MODEL_ID);
-            Model model = ModelCache.getInstance().get(modelId);
-            if (model.getModelBlob() == null) {
-                throw new RuntimeException(String.format("There is no trained model with id \"%s\"", modelId));
-            }
-            vectorDataType = model.getModelMetadata().getVectorDataType();
-            pair = KNNCodecUtil.getPair(values, getVectorTransfer(vectorDataType));
-            indexCreator = () -> createKNNIndexFromTemplate(model, pair, knnEngine, indexPath);
-        } else {
-            // get vector data type from field attributes or provide default value
-            vectorDataType = VectorDataType.get(
-                fieldAttributes.getOrDefault(KNNConstants.VECTOR_DATA_TYPE_FIELD, VectorDataType.DEFAULT.getValue())
-            );
-            pair = KNNCodecUtil.getPair(values, getVectorTransfer(vectorDataType));
-            indexCreator = () -> createKNNIndexFromScratch(field, pair, knnEngine, indexPath);
-        }
-
-        // Skip index creation if no vectors or docs in segment
-        if (pair.getVectorAddress() == 0 || pair.docs.length == 0) {
-            logger.info("Skipping engine index creation as there are no vectors or docs in the segment");
-            return;
-        }
-
-        long arraySize = calculateArraySize(pair.docs.length, pair.getDimension(), vectorDataType);
+    public void addKNNBinaryField(FieldInfo field, DocValuesProducer valuesProducer, boolean isMerge) throws IOException {
+        final VectorDataType vectorDataType = extractVectorDataType(field);
+        final KNNVectorValues<?> knnVectorValues = KNNVectorValuesFactory.getVectorValues(vectorDataType, valuesProducer.getBinary(field));
```

Comment:
How is this supposed to know the type without the wildcard use? What am I missing?

Reply:
I also want to know how we can avoid this. because the underline VectorValues class uses generic/ or we write a wrapper over the primitive arrays. But then in that case VectorTransfer classes has to figure out what is that generics.